File: wallet/common/hpke/hpke.go
// Package hpke implements Hybrid Public Key Encryption (RFC 9180) in base mode
// together with its JOSE binding, "Use of Hybrid Public Key Encryption (HPKE)
// with JSON Web Encryption (JWE)" (draft-ietf-jose-hpke-encrypt).
//
// Only the Integrated Encryption key management mode is implemented, which is
// the mode OpenID4VP 1.1 Section 8.3.1 builds on when encrypting Authorization
// Responses. In that mode the JWE Encrypted Key carries the HPKE encapsulated
// secret and no separate content encryption algorithm ("enc") is involved.
package hpke

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdh"
	"crypto/hkdf"
	"crypto/rand"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/binary"
	"fmt"
	"hash"

	"golang.org/x/crypto/chacha20poly1305"
)

// hpkeVersion is the version label that prefixes every labeled KDF input
// (RFC 9180 Section 4).
const hpkeVersion = "HPKE-v1"

// modeBase is the HPKE mode used by JOSE HPKE: no pre-shared key and no sender
// authentication (RFC 9180 Section 5.1).
const modeBase byte = 0x00

// KEM identifies an HPKE Key Encapsulation Mechanism from the IANA HPKE
// registry. X448 based KEMs are deliberately absent: the Go standard library
// offers no X448 implementation.
type KEM uint16

const (
	// KEMP256HKDFSHA256 is DHKEM(P-256, HKDF-SHA256).
	KEMP256HKDFSHA256 KEM = 0x0010
	// KEMP384HKDFSHA384 is DHKEM(P-384, HKDF-SHA384).
	KEMP384HKDFSHA384 KEM = 0x0011
	// KEMP521HKDFSHA512 is DHKEM(P-521, HKDF-SHA512).
	KEMP521HKDFSHA512 KEM = 0x0012
	// KEMX25519HKDFSHA256 is DHKEM(X25519, HKDF-SHA256).
	KEMX25519HKDFSHA256 KEM = 0x0020
)

// KDF identifies an HPKE Key Derivation Function from the IANA HPKE registry.
type KDF uint16

const (
	// KDFHKDFSHA256 is HKDF-SHA256.
	KDFHKDFSHA256 KDF = 0x0001
	// KDFHKDFSHA384 is HKDF-SHA384.
	KDFHKDFSHA384 KDF = 0x0002
	// KDFHKDFSHA512 is HKDF-SHA512.
	KDFHKDFSHA512 KDF = 0x0003
)

// AEAD identifies an HPKE AEAD from the IANA HPKE registry.
type AEAD uint16

const (
	// AEADAES128GCM is AES-128-GCM.
	AEADAES128GCM AEAD = 0x0001
	// AEADAES256GCM is AES-256-GCM.
	AEADAES256GCM AEAD = 0x0002
	// AEADChaCha20Poly1305 is ChaCha20Poly1305.
	AEADChaCha20Poly1305 AEAD = 0x0003
)

// Suite is an HPKE ciphersuite: the KEM, KDF and AEAD triple that both parties
// agree on out of band. In JOSE HPKE the "alg" header parameter names the suite.
type Suite struct {
	KEM  KEM
	KDF  KDF
	AEAD AEAD
}

// Curve returns the elliptic curve the suite's KEM operates on. It is exported
// so callers can check a recipient key against the suite before encrypting.
func (s Suite) Curve() (ecdh.Curve, error) {
	return s.KEM.curve()
}

// Seal performs a single-shot HPKE encryption in base mode (RFC 9180 Section
// 6.1). It returns the encapsulated secret and the ciphertext.
func (s Suite) Seal(recipient *ecdh.PublicKey, info, aad, plaintext []byte) (enc, ciphertext []byte, err error) {
	if recipient == nil {
		return nil, nil, fmt.Errorf("hpke: recipient public key is nil")
	}

	shared, enc, err := s.encap(recipient)
	if err != nil {
		return nil, nil, err
	}

	aead, nonce, err := s.aeadForContext(shared, info)
	if err != nil {
		return nil, nil, err
	}

	return enc, aead.Seal(nil, nonce, plaintext, aad), nil
}

// Open reverses Seal. The info and aad values must match the ones the sender
// used; anything else fails authentication rather than returning wrong data.
func (s Suite) Open(recipient *ecdh.PrivateKey, enc, info, aad, ciphertext []byte) ([]byte, error) {
	if recipient == nil {
		return nil, fmt.Errorf("hpke: recipient private key is nil")
	}

	shared, err := s.decap(recipient, enc)
	if err != nil {
		return nil, err
	}

	aead, nonce, err := s.aeadForContext(shared, info)
	if err != nil {
		return nil, err
	}

	plaintext, err := aead.Open(nil, nonce, ciphertext, aad)
	if err != nil {
		return nil, fmt.Errorf("hpke: decryption failed: %w", err)
	}
	return plaintext, nil
}

// encap implements DHKEM Encap (RFC 9180 Section 4.1).
func (s Suite) encap(pkR *ecdh.PublicKey) (shared, enc []byte, err error) {
	curve, err := s.KEM.curve()
	if err != nil {
		return nil, nil, err
	}
	if pkR.Curve() != curve {
		return nil, nil, fmt.Errorf("hpke: recipient key is on a different curve than KEM 0x%04x expects", uint16(s.KEM))
	}

	skE, err := curve.GenerateKey(rand.Reader)
	if err != nil {
		return nil, nil, fmt.Errorf("hpke: failed to generate ephemeral key: %w", err)
	}

	dh, err := skE.ECDH(pkR)
	if err != nil {
		return nil, nil, fmt.Errorf("hpke: key agreement failed: %w", err)
	}

	enc = skE.PublicKey().Bytes()
	shared, err = s.KEM.extractAndExpand(dh, concat(enc, pkR.Bytes()))
	if err != nil {
		return nil, nil, err
	}
	return shared, enc, nil
}

// decap implements DHKEM Decap (RFC 9180 Section 4.1).
func (s Suite) decap(skR *ecdh.PrivateKey, enc []byte) ([]byte, error) {
	curve, err := s.KEM.curve()
	if err != nil {
		return nil, err
	}
	if skR.Curve() != curve {
		return nil, fmt.Errorf("hpke: recipient key is on a different curve than KEM 0x%04x expects", uint16(s.KEM))
	}

	pkE, err := curve.NewPublicKey(enc)
	if err != nil {
		return nil, fmt.Errorf("hpke: invalid encapsulated secret: %w", err)
	}

	dh, err := skR.ECDH(pkE)
	if err != nil {
		return nil, fmt.Errorf("hpke: key agreement failed: %w", err)
	}

	return s.KEM.extractAndExpand(dh, concat(enc, skR.PublicKey().Bytes()))
}

// aeadForContext runs the base-mode key schedule (RFC 9180 Section 5.1) and
// returns the AEAD together with the nonce for sequence number 0. Single-shot
// encryption never advances the sequence number, so the base nonce is used
// unchanged.
func (s Suite) aeadForContext(shared, info []byte) (cipher.AEAD, []byte, error) {
	kdf, err := s.KDF.params()
	if err != nil {
		return nil, nil, err
	}
	aeadParams, err := s.AEAD.params()
	if err != nil {
		return nil, nil, err
	}

	suiteID := s.suiteID()

	// An empty psk_id and an empty psk are what mode_base prescribes.
	pskIDHash, err := labeledExtract(kdf, suiteID, nil, "psk_id_hash", nil)
	if err != nil {
		return nil, nil, err
	}
	infoHash, err := labeledExtract(kdf, suiteID, nil, "info_hash", info)
	if err != nil {
		return nil, nil, err
	}
	keyScheduleContext := concat([]byte{modeBase}, pskIDHash, infoHash)

	secret, err := labeledExtract(kdf, suiteID, shared, "secret", nil)
	if err != nil {
		return nil, nil, err
	}
	key, err := labeledExpand(kdf, suiteID, secret, "key", keyScheduleContext, aeadParams.keySize)
	if err != nil {
		return nil, nil, err
	}
	baseNonce, err := labeledExpand(kdf, suiteID, secret, "base_nonce", keyScheduleContext, aeadParams.nonceSize)
	if err != nil {
		return nil, nil, err
	}

	aead, err := aeadParams.new(key)
	if err != nil {
		return nil, nil, err
	}
	return aead, baseNonce, nil
}

// suiteID is the ciphersuite identifier mixed into the key schedule
// (RFC 9180 Section 5.1).
func (s Suite) suiteID() []byte {
	id := make([]byte, 0, 10)
	id = append(id, "HPKE"...)
	id = binary.BigEndian.AppendUint16(id, uint16(s.KEM))
	id = binary.BigEndian.AppendUint16(id, uint16(s.KDF))
	return binary.BigEndian.AppendUint16(id, uint16(s.AEAD))
}

// curve maps a KEM to the ecdh curve it performs Diffie-Hellman on.
func (k KEM) curve() (ecdh.Curve, error) {
	switch k {
	case KEMP256HKDFSHA256:
		return ecdh.P256(), nil
	case KEMP384HKDFSHA384:
		return ecdh.P384(), nil
	case KEMP521HKDFSHA512:
		return ecdh.P521(), nil
	case KEMX25519HKDFSHA256:
		return ecdh.X25519(), nil
	default:
		return nil, fmt.Errorf("hpke: unsupported KEM 0x%04x", uint16(k))
	}
}

// kdf returns the KDF that is baked into the KEM. It is independent of the
// suite's KDF, which is only used by the key schedule.
func (k KEM) kdf() (KDF, error) {
	switch k {
	case KEMP256HKDFSHA256, KEMX25519HKDFSHA256:
		return KDFHKDFSHA256, nil
	case KEMP384HKDFSHA384:
		return KDFHKDFSHA384, nil
	case KEMP521HKDFSHA512:
		return KDFHKDFSHA512, nil
	default:
		return 0, fmt.Errorf("hpke: unsupported KEM 0x%04x", uint16(k))
	}
}

// secretSize is Nsecret, the length of the KEM shared secret.
func (k KEM) secretSize() (int, error) {
	switch k {
	case KEMP256HKDFSHA256, KEMX25519HKDFSHA256:
		return 32, nil
	case KEMP384HKDFSHA384:
		return 48, nil
	case KEMP521HKDFSHA512:
		return 64, nil
	default:
		return 0, fmt.Errorf("hpke: unsupported KEM 0x%04x", uint16(k))
	}
}

// extractAndExpand derives the KEM shared secret (RFC 9180 Section 4.1).
func (k KEM) extractAndExpand(dh, kemContext []byte) ([]byte, error) {
	kdfID, err := k.kdf()
	if err != nil {
		return nil, err
	}
	kdf, err := kdfID.params()
	if err != nil {
		return nil, err
	}
	secretSize, err := k.secretSize()
	if err != nil {
		return nil, err
	}

	suiteID := binary.BigEndian.AppendUint16([]byte("KEM"), uint16(k))

	eaePRK, err := labeledExtract(kdf, suiteID, nil, "eae_prk", dh)
	if err != nil {
		return nil, err
	}
	return labeledExpand(kdf, suiteID, eaePRK, "shared_secret", kemContext, secretSize)
}

type kdfParams struct {
	newHash func() hash.Hash
}

func (k KDF) params() (kdfParams, error) {
	switch k {
	case KDFHKDFSHA256:
		return kdfParams{newHash: sha256.New}, nil
	case KDFHKDFSHA384:
		return kdfParams{newHash: sha512.New384}, nil
	case KDFHKDFSHA512:
		return kdfParams{newHash: sha512.New}, nil
	default:
		return kdfParams{}, fmt.Errorf("hpke: unsupported KDF 0x%04x", uint16(k))
	}
}

type aeadParams struct {
	keySize   int
	nonceSize int
	new       func(key []byte) (cipher.AEAD, error)
}

func (a AEAD) params() (aeadParams, error) {
	switch a {
	case AEADAES128GCM:
		return aeadParams{keySize: 16, nonceSize: 12, new: newAESGCM}, nil
	case AEADAES256GCM:
		return aeadParams{keySize: 32, nonceSize: 12, new: newAESGCM}, nil
	case AEADChaCha20Poly1305:
		return aeadParams{keySize: 32, nonceSize: 12, new: newChaCha20Poly1305}, nil
	default:
		return aeadParams{}, fmt.Errorf("hpke: unsupported AEAD 0x%04x", uint16(a))
	}
}

func newAESGCM(key []byte) (cipher.AEAD, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("hpke: failed to create AES cipher: %w", err)
	}
	aead, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("hpke: failed to create GCM: %w", err)
	}
	return aead, nil
}

func newChaCha20Poly1305(key []byte) (cipher.AEAD, error) {
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return nil, fmt.Errorf("hpke: failed to create ChaCha20Poly1305: %w", err)
	}
	return aead, nil
}

// labeledExtract is LabeledExtract from RFC 9180 Section 4.
func labeledExtract(kdf kdfParams, suiteID, salt []byte, label string, ikm []byte) ([]byte, error) {
	labeledIKM := concat([]byte(hpkeVersion), suiteID, []byte(label), ikm)
	prk, err := hkdf.Extract(kdf.newHash, labeledIKM, salt)
	if err != nil {
		return nil, fmt.Errorf("hpke: HKDF extract failed: %w", err)
	}
	return prk, nil
}

// labeledExpand is LabeledExpand from RFC 9180 Section 4.
func labeledExpand(kdf kdfParams, suiteID, prk []byte, label string, info []byte, length int) ([]byte, error) {
	if length < 0 || length > 0xFFFF {
		return nil, fmt.Errorf("hpke: invalid expand length %d", length)
	}
	labeledInfo := binary.BigEndian.AppendUint16(nil, uint16(length))
	labeledInfo = concat(labeledInfo, []byte(hpkeVersion), suiteID, []byte(label), info)

	out, err := hkdf.Expand(kdf.newHash, prk, string(labeledInfo), length)
	if err != nil {
		return nil, fmt.Errorf("hpke: HKDF expand failed: %w", err)
	}
	return out, nil
}

func concat(parts ...[]byte) []byte {
	size := 0
	for _, part := range parts {
		size += len(part)
	}
	out := make([]byte, 0, size)
	for _, part := range parts {
		out = append(out, part...)
	}
	return out
}

File: issuer+verifier/src/hpke.ts
import {
  CipherGCMTypes,
  createCipheriv,
  createDecipheriv,
  createHmac,
  createPrivateKey,
  createPublicKey,
  diffieHellman,
  generateKeyPairSync,
  KeyObject,
} from 'node:crypto'
import { err } from './errors/vcknots.error'
import { Jwk } from './jwk.type'

/**
 * Hybrid Public Key Encryption (RFC 9180) in base mode: no pre-shared key and
 * no sender authentication.
 *
 * This is the primitive behind the JOSE HPKE binding that OID4VP 1.1
 * Section 8.3.1 uses to encrypt Authorization Responses. DHKEM(X448) is absent
 * even though Node.js could perform it: the Go wallet in this repository has no
 * X448 to pair with, so the suite is left out of both sides together.
 */

/** The version label that prefixes every labeled KDF input (RFC 9180 Section 4). */
const HPKE_VERSION = 'HPKE-v1'

/** mode_base: no PSK, no sender authentication (RFC 9180 Section 5.1). */
const MODE_BASE = 0x00

/** KEM identifiers from the IANA HPKE registry. */
export const Kem = {
  P256_HKDF_SHA256: 0x0010,
  P384_HKDF_SHA384: 0x0011,
  P521_HKDF_SHA512: 0x0012,
  X25519_HKDF_SHA256: 0x0020,
} as const
export type KemId = (typeof Kem)[keyof typeof Kem]

/** KDF identifiers from the IANA HPKE registry. */
export const Kdf = {
  HKDF_SHA256: 0x0001,
  HKDF_SHA384: 0x0002,
  HKDF_SHA512: 0x0003,
} as const
export type KdfId = (typeof Kdf)[keyof typeof Kdf]

/** AEAD identifiers from the IANA HPKE registry. */
export const Aead = {
  AES_128_GCM: 0x0001,
  AES_256_GCM: 0x0002,
  CHACHA20_POLY1305: 0x0003,
} as const
export type AeadId = (typeof Aead)[keyof typeof Aead]

/** An HPKE ciphersuite. In JOSE HPKE the `alg` header parameter names one. */
export type HpkeSuite = {
  kem: KemId
  kdf: KdfId
  aead: AeadId
}

type KemParams = {
  /** The KDF baked into the KEM, independent of the suite's KDF. */
  kdf: KdfId
  /** Nsecret, the length of the KEM shared secret. */
  secretLength: number
  /** The JWK key type and curve the KEM operates on. */
  kty: 'EC' | 'OKP'
  crv: string
  /** The byte length of a single coordinate on the curve. */
  coordinateLength: number
  /** The Node.js key generation parameters. */
  generate: () => { publicKey: KeyObject; privateKey: KeyObject }
}

const kemParams: Record<KemId, KemParams> = {
  [Kem.P256_HKDF_SHA256]: {
    kdf: Kdf.HKDF_SHA256,
    secretLength: 32,
    kty: 'EC',
    crv: 'P-256',
    coordinateLength: 32,
    generate: () => generateKeyPairSync('ec', { namedCurve: 'prime256v1' }),
  },
  [Kem.P384_HKDF_SHA384]: {
    kdf: Kdf.HKDF_SHA384,
    secretLength: 48,
    kty: 'EC',
    crv: 'P-384',
    coordinateLength: 48,
    generate: () => generateKeyPairSync('ec', { namedCurve: 'secp384r1' }),
  },
  [Kem.P521_HKDF_SHA512]: {
    kdf: Kdf.HKDF_SHA512,
    secretLength: 64,
    kty: 'EC',
    crv: 'P-521',
    coordinateLength: 66,
    generate: () => generateKeyPairSync('ec', { namedCurve: 'secp521r1' }),
  },
  [Kem.X25519_HKDF_SHA256]: {
    kdf: Kdf.HKDF_SHA256,
    secretLength: 32,
    kty: 'OKP',
    crv: 'X25519',
    coordinateLength: 32,
    generate: () => generateKeyPairSync('x25519'),
  },
}

const kdfHash: Record<KdfId, string> = {
  [Kdf.HKDF_SHA256]: 'sha256',
  [Kdf.HKDF_SHA384]: 'sha384',
  [Kdf.HKDF_SHA512]: 'sha512',
}

type AeadParams = {
  /**
   * The Node.js cipher name. It is typed as the GCM family because all three
   * AEADs expose the same authenticated-encryption interface, and the narrower
   * `createCipheriv` overloads that accept `authTagLength` are keyed on it.
   */
  cipher: CipherGCMTypes
  keyLength: number
  nonceLength: number
  tagLength: number
}

const aeadParams: Record<AeadId, AeadParams> = {
  [Aead.AES_128_GCM]: { cipher: 'aes-128-gcm', keyLength: 16, nonceLength: 12, tagLength: 16 },
  [Aead.AES_256_GCM]: { cipher: 'aes-256-gcm', keyLength: 32, nonceLength: 12, tagLength: 16 },
  [Aead.CHACHA20_POLY1305]: {
    cipher: 'chacha20-poly1305' as CipherGCMTypes,
    keyLength: 32,
    nonceLength: 12,
    tagLength: 16,
  },
}

/** The JWK key type and curve a KEM expects its keys to use. */
export const kemKeyType = (kem: KemId): { kty: 'EC' | 'OKP'; crv: string } => {
  const params = kemParams[kem]
  return { kty: params.kty, crv: params.crv }
}

/** Generates a key pair for a KEM and returns it as a public and a private JWK. */
export const generateKemKeyPair = (kem: KemId): { publicKey: Jwk; privateKey: Jwk } => {
  const { publicKey, privateKey } = kemParams[kem].generate()
  return {
    publicKey: publicKey.export({ format: 'jwk' }) as Jwk,
    privateKey: privateKey.export({ format: 'jwk' }) as Jwk,
  }
}

/**
 * Encrypts plaintext to a recipient in a single shot (RFC 9180 Section 6.1),
 * returning the encapsulated secret alongside the ciphertext.
 */
export const seal = (
  suite: HpkeSuite,
  recipientPublicKey: Jwk,
  info: Uint8Array,
  aad: Uint8Array,
  plaintext: Uint8Array
): { enc: Buffer; ciphertext: Buffer } => {
  const { sharedSecret, enc } = encapsulate(suite.kem, recipientPublicKey)
  const { key, nonce } = keySchedule(suite, sharedSecret, info)
  const params = aeadParams[suite.aead]

  const cipher = createCipheriv(params.cipher, key, nonce, {
    authTagLength: params.tagLength,
  })
  cipher.setAAD(Buffer.from(aad))
  const body = Buffer.concat([cipher.update(Buffer.from(plaintext)), cipher.final()])

  return { enc, ciphertext: Buffer.concat([body, cipher.getAuthTag()]) }
}

/**
 * Reverses {@link seal}. The info and aad values must match the ones the sender
 * used; anything else fails authentication rather than returning wrong data.
 */
export const open = (
  suite: HpkeSuite,
  recipientPrivateKey: Jwk,
  enc: Uint8Array,
  info: Uint8Array,
  aad: Uint8Array,
  ciphertext: Uint8Array
): Buffer => {
  const params = aeadParams[suite.aead]
  if (ciphertext.length < params.tagLength) {
    throw err('invalid_encryption_parameters', {
      message: 'The HPKE ciphertext is shorter than its authentication tag.',
    })
  }

  const sharedSecret = decapsulate(suite.kem, recipientPrivateKey, enc)
  const { key, nonce } = keySchedule(suite, sharedSecret, info)

  const body = Buffer.from(ciphertext.subarray(0, ciphertext.length - params.tagLength))
  const tag = Buffer.from(ciphertext.subarray(ciphertext.length - params.tagLength))

  const decipher = createDecipheriv(params.cipher, key, nonce, {
    authTagLength: params.tagLength,
  })
  decipher.setAAD(Buffer.from(aad))
  decipher.setAuthTag(tag)

  try {
    return Buffer.concat([decipher.update(body), decipher.final()])
  } catch (_error) {
    throw err('invalid_encryption_parameters', {
      message:
        'HPKE decryption failed. The key, the ciphertext or the bound context does not match.',
    })
  }
}

/** DHKEM Encap (RFC 9180 Section 4.1). */
const encapsulate = (kem: KemId, recipientPublicKey: Jwk) => {
  const params = kemParams[kem]
  const recipient = importPublicKey(kem, recipientPublicKey)

  const ephemeral = params.generate()
  const dh = diffieHellman({ privateKey: ephemeral.privateKey, publicKey: recipient })

  const enc = serializePublicKey(kem, ephemeral.publicKey.export({ format: 'jwk' }) as Jwk)
  const kemContext = Buffer.concat([enc, serializePublicKey(kem, recipientPublicKey)])

  return { sharedSecret: extractAndExpand(kem, dh, kemContext), enc }
}

/** DHKEM Decap (RFC 9180 Section 4.1). */
const decapsulate = (kem: KemId, recipientPrivateKey: Jwk, enc: Uint8Array) => {
  const recipient = importPrivateKey(kem, recipientPrivateKey)
  const ephemeral = deserializePublicKey(kem, enc)

  const dh = diffieHellman({ privateKey: recipient, publicKey: ephemeral })
  const kemContext = Buffer.concat([
    Buffer.from(enc),
    serializePublicKey(kem, publicPartOf(recipientPrivateKey)),
  ])

  return extractAndExpand(kem, dh, kemContext)
}

/** ExtractAndExpand, which derives the KEM shared secret (RFC 9180 Section 4.1). */
const extractAndExpand = (kem: KemId, dh: Buffer, kemContext: Buffer): Buffer => {
  const params = kemParams[kem]
  const suiteId = Buffer.concat([Buffer.from('KEM'), uint16(kem)])

  const eaePrk = labeledExtract(params.kdf, suiteId, Buffer.alloc(0), 'eae_prk', dh)
  return labeledExpand(
    params.kdf,
    suiteId,
    eaePrk,
    'shared_secret',
    kemContext,
    params.secretLength
  )
}

/**
 * The base-mode key schedule (RFC 9180 Section 5.1). Single-shot encryption
 * never advances the sequence number, so the base nonce is the nonce.
 */
const keySchedule = (suite: HpkeSuite, sharedSecret: Buffer, info: Uint8Array) => {
  const aead = aeadParams[suite.aead]
  const suiteId = Buffer.concat([
    Buffer.from('HPKE'),
    uint16(suite.kem),
    uint16(suite.kdf),
    uint16(suite.aead),
  ])

  const empty = Buffer.alloc(0)
  // mode_base prescribes an empty psk and an empty psk_id.
  const pskIdHash = labeledExtract(suite.kdf, suiteId, empty, 'psk_id_hash', empty)
  const infoHash = labeledExtract(suite.kdf, suiteId, empty, 'info_hash', Buffer.from(info))
  const context = Buffer.concat([Buffer.from([MODE_BASE]), pskIdHash, infoHash])

  const secret = labeledExtract(suite.kdf, suiteId, sharedSecret, 'secret', empty)

  return {
    key: labeledExpand(suite.kdf, suiteId, secret, 'key', context, aead.keyLength),
    nonce: labeledExpand(suite.kdf, suiteId, secret, 'base_nonce', context, aead.nonceLength),
  }
}

/** LabeledExtract (RFC 9180 Section 4). */
const labeledExtract = (
  kdf: KdfId,
  suiteId: Buffer,
  salt: Buffer,
  label: string,
  ikm: Buffer
): Buffer => {
  const labeledIkm = Buffer.concat([Buffer.from(HPKE_VERSION), suiteId, Buffer.from(label), ikm])
  return hkdfExtract(kdfHash[kdf], salt, labeledIkm)
}

/** LabeledExpand (RFC 9180 Section 4). */
const labeledExpand = (
  kdf: KdfId,
  suiteId: Buffer,
  prk: Buffer,
  label: string,
  info: Buffer,
  length: number
): Buffer => {
  const labeledInfo = Buffer.concat([
    uint16(length),
    Buffer.from(HPKE_VERSION),
    suiteId,
    Buffer.from(label),
    info,
  ])
  return hkdfExpand(kdfHash[kdf], prk, labeledInfo, length)
}

/**
 * HKDF-Extract (RFC 5869 Section 2.2). Node's `hkdfSync` fuses extract and
 * expand, but HPKE needs the pseudorandom key on its own.
 */
const hkdfExtract = (hash: string, salt: Buffer, ikm: Buffer): Buffer => {
  const hashLength = createHmac(hash, Buffer.alloc(0)).digest().length
  const key = salt.length > 0 ? salt : Buffer.alloc(hashLength)
  return createHmac(hash, key).update(ikm).digest()
}

/** HKDF-Expand (RFC 5869 Section 2.3). */
const hkdfExpand = (hash: string, prk: Buffer, info: Buffer, length: number): Buffer => {
  const hashLength = createHmac(hash, Buffer.alloc(0)).digest().length
  if (length > 255 * hashLength) {
    throw err('illegal_argument', { message: 'HKDF cannot expand to the requested length.' })
  }

  const blocks: Buffer[] = []
  let previous = Buffer.alloc(0)
  for (let counter = 1; Buffer.concat(blocks).length < length; counter++) {
    previous = createHmac(hash, prk)
      .update(Buffer.concat([previous, info, Buffer.from([counter])]))
      .digest()
    blocks.push(previous)
  }
  return Buffer.concat(blocks).subarray(0, length)
}

/** SerializePublicKey: the SEC 1 uncompressed point, or the raw X25519 key. */
const serializePublicKey = (kem: KemId, jwk: Jwk): Buffer => {
  const params = kemParams[kem]
  const x = decodeCoordinate(jwk.x, 'x', params.coordinateLength)

  if (params.kty === 'OKP') return x
  return Buffer.concat([
    Buffer.from([0x04]),
    x,
    decodeCoordinate(jwk.y, 'y', params.coordinateLength),
  ])
}

/** DeserializePublicKey, the inverse of {@link serializePublicKey}. */
const deserializePublicKey = (kem: KemId, encoded: Uint8Array): KeyObject => {
  const params = kemParams[kem]
  const bytes = Buffer.from(encoded)
  const encode = (value: Buffer) => value.toString('base64url')

  if (params.kty === 'OKP') {
    if (bytes.length !== params.coordinateLength) {
      throw err('invalid_encryption_parameters', {
        message: `The encapsulated secret is ${bytes.length} bytes, expected ${params.coordinateLength}.`,
      })
    }
    return importPublicKey(kem, { kty: 'OKP', crv: params.crv, x: encode(bytes) })
  }

  const expected = 1 + params.coordinateLength * 2
  if (bytes.length !== expected || bytes[0] !== 0x04) {
    throw err('invalid_encryption_parameters', {
      message: 'The encapsulated secret is not an uncompressed elliptic curve point.',
    })
  }
  return importPublicKey(kem, {
    kty: 'EC',
    crv: params.crv,
    x: encode(bytes.subarray(1, 1 + params.coordinateLength)),
    y: encode(bytes.subarray(1 + params.coordinateLength)),
  })
}

const importPublicKey = (kem: KemId, jwk: Jwk): KeyObject => {
  assertKeyMatchesKem(kem, jwk)
  try {
    return createPublicKey({ key: jwk as Record<string, unknown>, format: 'jwk' })
  } catch (error) {
    throw err('invalid_encryption_parameters', {
      message: `The HPKE public key could not be imported: ${error}`,
    })
  }
}

const importPrivateKey = (kem: KemId, jwk: Jwk): KeyObject => {
  assertKeyMatchesKem(kem, jwk)
  if (typeof jwk.d !== 'string' || jwk.d === '') {
    throw err('invalid_encryption_parameters', { message: 'The HPKE key holds no private part.' })
  }
  try {
    return createPrivateKey({ key: jwk as Record<string, unknown>, format: 'jwk' })
  } catch (error) {
    throw err('invalid_encryption_parameters', {
      message: `The HPKE private key could not be imported: ${error}`,
    })
  }
}

const assertKeyMatchesKem = (kem: KemId, jwk: Jwk) => {
  const params = kemParams[kem]
  if (jwk.kty !== params.kty || jwk.crv !== params.crv) {
    throw err('invalid_encryption_parameters', {
      message: `The HPKE key is ${jwk.kty}/${jwk.crv}, but the algorithm needs ${params.kty}/${params.crv}.`,
    })
  }
}

/** Drops the private members of a JWK, leaving the public key. */
const publicPartOf = (jwk: Jwk): Jwk => {
  const { d: _d, ...publicKey } = jwk
  return publicKey
}

const decodeCoordinate = (value: unknown, name: string, length: number): Buffer => {
  if (typeof value !== 'string' || value === '') {
    throw err('invalid_encryption_parameters', { message: `The JWK member "${name}" is missing.` })
  }
  const decoded = Buffer.from(value, 'base64url')
  if (decoded.length !== length) {
    throw err('invalid_encryption_parameters', {
      message: `The JWK member "${name}" is ${decoded.length} bytes, expected ${length}.`,
    })
  }
  return decoded
}

const uint16 = (value: number): Buffer => {
  const encoded = Buffer.alloc(2)
  encoded.writeUInt16BE(value)
  return encoded
}
